Feat/book stock

esterinac opened the pull request:
1. Hapus detail stok buku di view buku
    Menghilangkan data stok buku di view_book karena data stok akan ditampilkan di view book stock.
![20210408_hapus detail data stok buku di view buku](https://user-images.githubusercontent.com/66486312/114002012-aec33380-9886-11eb-9c9e-4179cdfab59b.JPG)

2. Migrasi data
    - Update table book_stock:
      - Menambahkan kolom library_present untuk menyimpan data stok buku di perpustakaan (secara keseluruhan)
      - Menambahkan kolom showroom_present untuk menyimpan data stok buku di showroom
      - Menambahkan kolom book_location untuk menyimpan data lokasi buku di gudang (Misal buku X ada di rak buku nomor berapa)
      - Menghapus kolom yang tidak digunakan di table book_stock
      - Contoh data book_stock:
![20210408_contoh data book stock](https://user-images.githubusercontent.com/66486312/114002062-bc78b900-9886-11eb-8af9-7d83094e9a5b.JPG)
    - Membuat table book_stock_revision:
      - Table book_stock_revision dibuat untuk mencatat revisi/edit stok buku yang ada di gudang. Revisi stok dilakukan dengan tambah/kurang stok seperti pada fitur yang sebelumnya sudah ada di view_book.
      - Contoh data book_stock_revision:    
![20210408_contoh data book stock revision](https://user-images.githubusercontent.com/66486312/114002104-c4d0f400-9886-11eb-878d-296f971216bf.JPG)

3. Book stock
    - Index book stock (berisi informasi stok dan generate excel data stok buku)
![20210408_index stok gudang](https://user-images.githubusercontent.com/66486312/114002143-d0bcb600-9886-11eb-9cae-acca3b810c79.JPG)
![20210408_index stok gudang 1](https://user-images.githubusercontent.com/66486312/114002153-d1ede300-9886-11eb-8cae-a050da34e0ec.JPG)
    - Edit/revisi book stock
      - Data revisi book stock akan ditampilkan di halaman view book stock
![20210408_edit stok](https://user-images.githubusercontent.com/66486312/114002244-e3cf8600-9886-11eb-9218-b73449c1c0f7.png)

    - View book stock
      - Belum menampilkan detail stok perpustakaan dan chart transaksi buku    
![20210408_view stok gudang 1](https://user-images.githubusercontent.com/66486312/114002178-d74b2d80-9886-11eb-8527-231114371a32.png)
![20210408_view stok gudang 2](https://user-images.githubusercontent.com/66486312/114002193-d9ad8780-9886-11eb-8121-531946218177.JPG)

    - Delete book stock
![20210408_index stok gudang 2](https://user-images.githubusercontent.com/66486312/114002521-24c79a80-9887-11eb-90a1-7ae771f28f7b.JPG)

esterinac committed:
Remove book stock details on view_book

esterinac committed:
Update table book_stock & create table book_stock_revision

- Add column library_present for library stock in table book_stock
- Add column warehouse_present for warehouse stock in table book_stock
- Add column book_location for warehouse book (shelf) location in table book_stock
- Drop unneeded column in table book_stock
- Create book_stock_revision table for storing book stock edit/revision data

esterinac committed:
Book stock

read, edit, and delete book stock

esterinac committed:
Fix edit book_stock

bagaskarala commented on application/helpers/ugmpress_helper.php:
```diff
@@ -795,3 +795,11 @@ function strip_disallowed_char($string)
+
+function get_warehouse_stock(){
+    return [
+        '' => 'Semua',
+        '1' => '<= 50',
+        '2' => '> 50'
```
untuk data seperti ini, sebaiknya pakai string aja untuk key.nya, misal `below_50` gitu?

esterinac replied:
baik mas, key nya jadi up_to_50 sama above_50 ya mas

bagaskarala commented on application/modules/book_stock/controllers/Book_stock.php:
```diff
@@ -0,0 +1,226 @@
+<?php defined('BASEPATH') or exit('No direct script access allowed');
+use PhpOffice\PhpSpreadsheet\Spreadsheet;
+use PhpOffice\PhpSpreadsheet\Writer\Xlsx;
```
error ketika klik tombol excel
![image](https://user-images.githubusercontent.com/14112952/114138387-c4dcfc80-9937-11eb-9be9-1a95fb95f6f9.png)

esterinac replied:
di tempatku bisa mas, atau mungkin karena packages nya blm keinstall? mungkin coba 'composer install' atau 'composer require phpoffice/phpspreadsheet' mas


https://user-images.githubusercontent.com/66486312/114143878-e42b5800-993e-11eb-8d57-d039a8c792c0.mp4

bagaskarala commented on application/modules/book_stock/controllers/Book_stock.php:
```diff
@@ -0,0 +1,226 @@
+                $this->session->set_flashdata('warning', $this->lang->line('toast_data_not_available'));
+            }
+            else {
+                if ($operator=="+") $book_stock->warehouse_present += $quantity;
```
oiya untuk tipe operator sebaiknya jangan pake symbol ya, mungkin pake `add` dan `sub` lebih readable

bagaskarala replied:
mungkin kata `operator` juga kurang oke, bisa diganti ke `revision_type` atau apa gitu?

esterinac replied:
baik mas, berarti di file migrasi buat table book_stock_revision diganti juga ya mas? soalnya di table itu yg operator tipenya varchar(1) 
![image](https://user-images.githubusercontent.com/66486312/114141775-4040ad00-993c-11eb-930b-d9344c269ae1.png)

bagaskarala replied:
> baik mas, berarti di file migrasi buat table book_stock_revision diganti juga ya mas? soalnya di table itu yg operator tipenya varchar(1)
> ![image](https://user-images.githubusercontent.com/66486312/114141775-4040ad00-993c-11eb-930b-d9344c269ae1.png)

oh iya diubah juga brrti, disesuaikan

bagaskarala commented:
@esterinac ini kan mindahin fitur dari book detail, ke book stock. code yang di book dihapus aja kalau gak dipakai lagi, biar bersih.

`user_id` yang mengubah stok dicatat gak? di table gak ada kolomnya.

esterinac commented:
> @esterinac ini kan mindahin fitur dari book detail, ke book stock. code yang di book dihapus aja kalau gak dipakai lagi, biar bersih.
> 
> `user_id` yang mengubah stok dicatat gak? di table gak ada kolomnya.

baik mas, untuk yang sekarang ini belum mencatat user yang mengubah stok mas. mungkin nanti kalau dari pak wayan/staf ugm press ingin kami mencatat data user yang mengedit stok, akan kami tambahkan mas.

esterinac committed:
Fix key to string on get_warehouse_stock() (book stock filter)

esterinac committed:
Remove unused book_stock code on book

esterinac committed:
Fix naming

- Change column name and type operator (varchar(1)) to revision_type (varchar(3)) on table book_stock_revision
- Change input value in revision_type from +/- to add/sub

esterinac committed:
Fix leveling on book stock

- Create Warehouse_Sales_Controller for features which can only be accessed by admin_gudang & admin_pemasaran
- Fix level on sidebar book_stock

bagaskarala commented:
@esterinac Oke² sip

esterinac committed:
Fix book_stock_revision migration file name
